Assertion trips in SearchHits$Total#writeTo in CI

A CI failure https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/542/console has exposed the tripping of a new assertion when running in a mixed cluster. The assertion that trips is:

https://github.com/elastic/elasticsearch/blob/2d03eeb632d098f66ae0cefdbb5d28150a4ea942/server/src/main/java/org/elasticsearch/search/SearchHits.java#L313

When this assertion is hit, node-2 terminates in the test which results in the search failure due to the node being disconnected.

```
[2018-12-05T20:44:21,496][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [node-2] fatal error in thread [elasticsearch[node-2][search][T#21]], exiting
java.lang.AssertionError: null
	at org.elasticsearch.search.SearchHits$Total.writeTo(SearchHits.java:313) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.search.SearchHits.writeTo(SearchHits.java:213) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.search.fetch.FetchSearchResult.writeTo(FetchSearchResult.java:102) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.search.fetch.QueryFetchSearchResult.writeTo(QueryFetchSearchResult.java:101) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TcpTransport.buildMessage(TcpTransport.java:946) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TcpTransport.sendResponse(TcpTransport.java:899) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TcpTransport.sendResponse(TcpTransport.java:872) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TcpTransportChannel.sendResponse(TcpTransportChannel.java:70) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:54) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.HandledTransportAction$ChannelActionListener.onResponse(HandledTransportAction.java:93) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.action.support.HandledTransportAction$ChannelActionListener.onResponse(HandledTransportAction.java:77) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:357) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:353) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.search.SearchService$4.doRun(SearchService.java:1094) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:759) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:834) [?:?]
```

I have not been able to reproduce this, but the reproduce line is:

```
./gradlew :qa:mixed-cluster:v6.6.0#mixedClusterTestRunner \
  -Dtests.seed=D53E935D92AC375C \
  -Dtests.class=org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT \
  -Dtests.method="test {p0=indices.sort/10_basic/Index Sort}" \
  -Dtests.security.manager=true \
  -Dtests.locale=cs-CZ \
  -Dtests.timezone=America/Santiago \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

--- a/server/src/main/java/org/elasticsearch/common/lucene/Lucene.java
+++ b/server/src/main/java/org/elasticsearch/common/lucene/Lucene.java
@@ -298,7 +298,7 @@ public class Lucene {
         return false;
     }
 
-    private static TotalHits readTotalHits(StreamInput in) throws IOException {
+    public static TotalHits readTotalHits(StreamInput in) throws IOException {
         long totalHits = in.readVLong();
         TotalHits.Relation totalHitsRelation = TotalHits.Relation.EQUAL_TO;
         if (in.getVersion().onOrAfter(org.elasticsearch.Version.V_7_0_0)) {
@@ -418,7 +418,7 @@ public class Lucene {
 
     private static final Class<?> GEO_DISTANCE_SORT_TYPE_CLASS = LatLonDocValuesField.newDistanceSort("some_geo_field", 0, 0).getClass();
 
-    private static void writeTotalHits(StreamOutput out, TotalHits totalHits) throws IOException {
+    public static void writeTotalHits(StreamOutput out, TotalHits totalHits) throws IOException {
         out.writeVLong(totalHits.value);
         if (out.getVersion().onOrAfter(org.elasticsearch.Version.V_7_0_0)) {
             out.writeEnum(totalHits.relation);

--- a/server/src/main/java/org/elasticsearch/search/SearchHits.java
+++ b/server/src/main/java/org/elasticsearch/search/SearchHits.java
@@ -21,12 +21,12 @@ package org.elasticsearch.search;
 
 import org.apache.lucene.search.TotalHits;
 import org.apache.lucene.search.TotalHits.Relation;
-import org.elasticsearch.Version;
 import org.elasticsearch.common.Nullable;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.io.stream.Streamable;
 import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.lucene.Lucene;
 import org.elasticsearch.common.xcontent.ToXContentFragment;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentParser;
@@ -276,14 +276,7 @@ public final class SearchHits implements Streamable, ToXContentFragment, Iterabl
         final TotalHits in;
 
         Total(StreamInput in) throws IOException {
-            final long value = in.readVLong();
-            final Relation relation;
-            if (in.getVersion().onOrAfter(Version.V_7_0_0)) {
-                relation = in.readEnum(Relation.class);
-            } else {
-                relation = Relation.EQUAL_TO;
-            }
-            this.in = new TotalHits(value, relation);
+            this.in = Lucene.readTotalHits(in);
         }
 
         Total(TotalHits in) {
@@ -306,12 +299,7 @@ public final class SearchHits implements Streamable, ToXContentFragment, Iterabl
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
-            out.writeVLong(in.value);
-            if (out.getVersion().onOrAfter(Version.V_7_0_0)) {
-                out.writeEnum(in.relation);
-            } else {
-                assert in.relation == Relation.EQUAL_TO;
-            }
+            Lucene.writeTotalHits(out, in);
         }
 
         @Override

